vw/src/main/scala/com/microsoft/azure/synapse/ml/vw/VowpalWabbitCSETransformer.scala

```diff
@@ -0,0 +1,208 @@
+    with Wrappable
+    with ComplexParamsWritable {
+
+  import VowpalWabbitDSJsonTransformer._
+  import VowpalWabbitCSETransformer._
+
+  logClass()
+
+  def this() = this(Identifiable.randomUID("VowpalWabbitEvalTransformer"))
+
+  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)
+
+  val minImportanceWeight = new FloatParam(
+    this, "dsJsonColumn", "Cap importance weight at this lower bound. Defaults to 0.")
+
+  def getMinImportanceWeight: Float = $(minImportanceWeight)
+  def getMinImportanceWeight(value: Float): this.type = set(minImportanceWeight, value)
+
+  val maxImportanceWeight = new FloatParam(
+    this, "dsJsonColumn", "Cap importance weight at this upper bound. Defaults to 100.")
+
+  def getMaxImportanceWeight: Float = $(maxImportanceWeight)
+  def getMaxImportanceWeight(value: Float): this.type = set(maxImportanceWeight, value)
+
+  val metricsStratificationCols = new StringArrayParam(
+    this, "metricsStratificationCols", "Optional list of column names to stratify rewards by.")
+
+  def getMetricsStratificationCols: Array[String] = $(metricsStratificationCols)
+  def getMetricsStratificationCols(value: Array[String]): this.type = set(metricsStratificationCols, value)
+
+  setDefault(minImportanceWeight -> 0, maxImportanceWeight -> 100, metricsStratificationCols -> Array.empty)
+
+  // define reward independent metrics
+  val globalMetrics: Seq[Column] = Seq(
+    F.count("*").alias(ExampleCountName),
+    F.sum(F.when(F.expr(s"$ProbabilityPredictedColName > 0"), 1).otherwise(0))
+      .alias(ProbabilityPredictedNonZeroCount),
+    F.min("w").alias(MinimumImportanceWeight),
+    F.max("w").alias(MaximumImportanceWeight),
+    F.expr("avg(w)").alias(AverageImportanceWeight),
+    F.expr("avg(w * w)").alias(AverageSquaredImportanceWeight),
+    F.expr("max(w)/count(*)").alias(PropOfMaximumImportanceWeight),
+    F.expr("approx_percentile(w, array(0.25, 0.5, 0.75, 0.95))")
+      .alias(ImportanceWeightQuantiles))
+
+  case class RewardColumn(name: String, col: String, idx: Int) {
+    def minRewardCol: String = s"min_reward_$idx"
+
+    def maxRewardCol: String = s"max_reward_$idx"
+  }
+
+  def rewardColumns(schema: T.StructType): Seq[RewardColumn] =
+    schema(RewardsColName)
+      .dataType.asInstanceOf[T.StructType]
+      .fields
+      .zipWithIndex
+      .map({ case (rewardField, idx) =>
+        RewardColumn(rewardField.name, s"$RewardsColName.${rewardField.name}", idx)
+      })
+      .toSeq
+
+
+  def perRewardMetrics(rewardsCol: Seq[RewardColumn],
+                       minImportanceWeight: Float = 0,
+                       maxImportanceWeight: Float = 100): Seq[Column] =
+    rewardsCol
+      .map({ rewardCol => {
+        F.struct(
+          F.first(F.col(rewardCol.minRewardCol)).alias(MinReward),
+          F.first(F.col(rewardCol.maxRewardCol)).alias(MaxReward),
+          F.expr(s"snips($ProbabilityLoggedColName, ${rewardCol.col}, $ProbabilityPredictedColName, count)")
```

Comment:
nit: just use the functional API instead of string interpolation

Reply:
can you show me how to do this with custom functions? I couldn't figure... (the snips(...)) is my problem.

Reply:
done